lib/teleterm/services/userpreferences/userpreferences.go

```diff
@@ -0,0 +1,124 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package userpreferences
+
+import (
+	"context"
+
+	"github.com/gravitational/trace"
+
+	userpreferencesv1 "github.com/gravitational/teleport/api/gen/proto/go/userpreferences/v1"
+	api "github.com/gravitational/teleport/gen/proto/go/teleport/lib/teleterm/v1"
+)
+
+func Get(ctx context.Context, rootClient Client, leafClient Client) (*api.UserPreferences, error) {
+	rootPreferencesResponse, err := rootClient.GetUserPreferences(ctx, &userpreferencesv1.GetUserPreferencesRequest{})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	rootPreferences := rootPreferencesResponse.GetPreferences()
+
+	var leafPreferences *userpreferencesv1.UserPreferences
+	if leafClient != nil {
+		preferences, err := leafClient.GetUserPreferences(ctx, &userpreferencesv1.GetUserPreferencesRequest{})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		leafPreferences = preferences.GetPreferences()
+	}
+
+	clusterPreferences := rootPreferences.GetClusterPreferences()
+	if leafPreferences != nil {
+		clusterPreferences = leafPreferences.GetClusterPreferences()
+	}
+
+	return &api.UserPreferences{
+		UnifiedResourcePreferences: rootPreferences.UnifiedResourcePreferences,
+		ClusterPreferences:         clusterPreferences,
+	}, nil
+}
+
+// Update updates the preferences for a given user.
+// Only the properties that are set (cluster_preferences, unified_resource_preferences) will be updated.
+func Update(ctx context.Context, rootClient Client, leafClient Client, newPreferences *api.UserPreferences) (*api.UserPreferences, error) {
+	// We have to fetch the full user preferences struct and modify only
+	// the fields that change.
+	// Calling `UpsertUserPreferences` with only the modified values would reset
+	// the rest of the preferences.
+	rootPreferencesResponse, err := rootClient.GetUserPreferences(ctx, &userpreferencesv1.GetUserPreferencesRequest{})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	rootPreferences := rootPreferencesResponse.GetPreferences()
+
+	var leafPreferences *userpreferencesv1.UserPreferences
+	if leafClient != nil {
+		response, err := leafClient.GetUserPreferences(ctx, &userpreferencesv1.GetUserPreferencesRequest{})
+		if err != nil {
+			return nil, trace.Wrap(err)
+		}
+		leafPreferences = response.GetPreferences()
+	}
+
+	hasUnifiedResourcePreferencesForRoot := newPreferences.UnifiedResourcePreferences != nil
+	hasClusterPreferencesForRoot := newPreferences.ClusterPreferences != nil && leafPreferences == nil
+
+	if hasUnifiedResourcePreferencesForRoot || hasClusterPreferencesForRoot {
+		if hasUnifiedResourcePreferencesForRoot {
+			rootPreferences.UnifiedResourcePreferences = newPreferences.UnifiedResourcePreferences
+		}
+		if hasClusterPreferencesForRoot {
+			rootPreferences.ClusterPreferences = newPreferences.ClusterPreferences
+		}
```

Comment:
How sure are you that this code is able to seamlessly handle the following situations?

* Someone adds or removes a new field to or from `rootPreferences`.
* Someone adds or removes a new field to or from `UnifiedResourcePreferences` or `ClusterPreferences`.

I'm mostly thinking of cases where we'd unknowingly overwrite (or turn to default) some preferences without realizing that. My naive 2 minute thinking shows that there's probably no reason to suspect that this can happen but I haven't thought about it too much.

Reply:
I was wanted to write a comment about it, because I'm a little worried about it as well.

I think I will change this code to only update the fields in `UnifiedResourcePreferences` and `ClusterPreferences`. This will make sure that we don't overwrite anything by an accident.

>Someone adds or removes a new field to or from rootPreferences.

I think it is already handled, we update only the fields of `rootPreferences`.